server/gfmd/user.c

```diff
@@ -1929,99 +1933,80 @@ gfm_server_user_auth_set_and_modify(struct peer *peer,
-		   &username, &auth_user_id_type_str);
-	if (e != GFARM_ERR_NO_ERROR) {
-		gflog_debug(GFARM_MSG_UNFIXED,
-		    "%s request: %s", diag, gfarm_error_string(e));
-		return (e);
-	}
-	if (skip) {
-		free(username);
-		free(auth_user_id_type_str);
-		return (GFARM_ERR_NO_ERROR);
-	}
-
-	e = gfarm_auth_user_id_type_parse(auth_user_id_type_str,
+	e = gfarm_auth_user_id_type_from_name(auth_user_id_type_str,
 					  &auth_user_id_type);
 	if (e != GFARM_ERR_NO_ERROR) {
 		gflog_debug(GFARM_MSG_UNFIXED, "invalid auth_user_id_type");
-		return (e);
-	}
-
-	/* XXX FIXME too long giant lock */
-	giant_lock();
-
-	if ((e = rpc_name_with_tenant(peer, from_client,
-	    &name_with_tenant, &process, diag)) != GFARM_ERR_NO_ERROR) {
+	} else if (!from_client || user == NULL) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_1003460, "%s: %s", diag,
+			gfarm_error_string(e));
+	} else if ((process = peer_get_process(peer)) == NULL) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s): no process",
+			diag, peer_get_username(peer), peer_get_hostname(peer));
+	} else if ((tenant = process_get_tenant(process)) == NULL) {
+		e = GFARM_ERR_INTERNAL_ERROR;
+		gflog_error(GFARM_MSG_UNFIXED, "%s (%s@%s): no tenant: %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			gfarm_error_string(e));
+	} else if (!user_is_tenant_admin(user, tenant)) {
+		e = GFARM_ERR_OPERATION_NOT_PERMITTED;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s): %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			gfarm_error_string(e));
+	} else if ((u = user_is_super_admin(user) ?
+			user_tenant_lookup(username) :
+			user_lookup_in_tenant(username, tenant))
+			== NULL) {
+		e = GFARM_ERR_NO_SUCH_USER;
+		gflog_debug(GFARM_MSG_UNFIXED, "%s (%s@%s) %s: %s",
+			diag, peer_get_username(peer), peer_get_hostname(peer),
+			username, gfarm_error_string(e));
+	} else if ((e = user_info_verify(&u->ui, diag)) != GFARM_ERR_NO_ERROR) {
+		gflog_debug(GFARM_MSG_1003462,
+			"%s: user_info_verify: %s",
+			diag, gfarm_error_string(e));
+	} else if (gfarm_read_only_mode()) {
+		gflog_debug(GFARM_MSG_1005143, "%s (%s@%s) during read_only",
+			diag, peer_get_username(peer), peer_get_hostname(peer));
+		e = GFARM_ERR_READ_ONLY_FILE_SYSTEM;
+	} else if ((e = rpc_name_with_tenant(peer, from_client,
+			&name_with_tenant, &process, diag))
+			!= GFARM_ERR_NO_ERROR) {
 		/* nothing to do */
 	} else if ((u = user_tenant_lookup(username)) == NULL) {
 		e = GFARM_ERR_NO_SUCH_USER;
 	} else
 		e = GFARM_ERR_NO_ERROR;
 
 	if (e == GFARM_ERR_NO_ERROR) {
+
+		if (auth_user_id == NULL || strcmp(auth_user_id, "") == 0) {
+			struct db_user_auth_remove_arg arg = {
+				username,
+				auth_user_id_type_str
+			};
+			user_auth_id_remove(u, auth_user_id_type_str);
+
+			struct user_auth_key key = {
+				auth_user_id_type,
+				auth_user_id
+			};
+			gfarm_hash_purge(user_id_map, &key, sizeof(key));
```

Comment:
the updates to u->auth_user_id[] and user_id_map should be done in user_auth_id_remove().
and the return value fof user_auth_id_remove() should be returned as the RPC result.